### ISSUE TITLE
Support storage/retrieval/query of capabilities

### ISSUE THREAD

**Issue by autodidaddict** [wascc/gantry#13]
Gantry currently supports the storage of tokens for Operators, Actors, and Accounts, and it supports the storage of binaries for actors. I would like to consider the possibility of storing capability providers in Gantry as well, but there are some issues that need to be ironed out.

* Actors have embedded signed JWTs, which makes them tamper-proof and have verifiable provenance, capability providers today have none of these features (except for portable capability providers which are WASI modules)
* Native Capability Providers are just that, native. This means that they lack portability, and so we'd have to potentially store multiple architectures for each capability provider.
* Native Capability Providers do not have globally unique identifiers. They have a capability ID (e.g. `wascc:graphdb` or `wascc:logging`, which are the equivalent of namespaces), and they have a friendly name (used primarily for logging), but they don't have a globally unique ID. So, storing capability providers and querying them might be awkward, especially when you've got multiple providers supporting the same capability (e.g. the File System Provider and the S3 provider both implement the `wascc:blobstore` capability).
* Native Capability Providers do not have the concept of a revision. Actors have a monotonically increasing revision number, which lets us pull the "latest" image of an actor or pin to a fixed one. If we were to support that same functionality for native providers, we'd need to find a way to give them revision numbers/versions.

Given the above concerns, another question is - is it worth it for Gantry to support storage of capability providers today, or do we wait until WASI supports robust networking and we can code all of our capability providers as portable WASI modules? The obvious risk there is no one knows how long it'll be before that happens.

**Comment by autodidaddict** [comment]
One possible solution to the "not enough metadata" and "not enough security" on the capability providers is to create a wrapper archive that's basically a `.tgz` file (maybe with a custom extension like `.cpar` for capability provider archive). This archive could contain:

* A library file binary for each `target triple` that provider was built for
* A signed JWT, though this might need to contain one hash per target triple rather than a single hash like the existing wascap tokens have

This would then set up the list of tokens managed by gantry to:
* Operator
* Account
* Actor
* Capability Provider

The gantry client could then require the caller to supply a list of all of the library files for a given provider, the signing key and module key, and then the gantry client would create a signed JWT for that provider, produce an archive containing the JWT and the library files, and then upload it to gantry.

The gantry protocol could then be updated to support requesting capability providers _by target triple_ and revision number.

**Comment by autodidaddict** [comment]
This model could also support storing both native (`.cpar` archives) and portable (wasi .wasm actors) capability providers side-by-side, allowing us to maintain both until we no longer need native.

**Comment by autodidaddict** [comment]
Given how potentially unreliable it is to detect a triplet from an arbitrary file, the Gantry client will likely need to require the consumer to manually specify the triplet for each library, if we end up going the archive route.

**Comment by brooksmtownsend** [comment]
To me, it seems like this will require either substantial changes on Gantry's part to support, or it will result in a bad user experience trying to retrieve a capability provider. It goes against the portability story of wasm to have to say "You can do whatever you want with these actors! And you can use whatever HTTP server you like! But you do have to get this specific one for your machine and change it if you're moving your code to a different place."

As long as I'm understanding the problem correctly, even with a correct Gantry setup a user would still have to be careful to grab the correct provider from their `triplet`. With the inclusion of statically linked capabilities, an easy way to use different capabilities regardless of architecture is available.

I think this brings up a good point though, that capability providers do not have a revision number and you're essentially always using "latest" and that could cause issues down the line with new refactors.

**Comment by autodidaddict** [comment]
I think the `user` exposure to this might not be as much as you think. Let's say you start a waSCC host and you just supply a triplet for it (`aarch64-blah-blah` on a raspberry Pi, for example). Then that host starts up with no capabilities and no actors. Developer code in a custom host then calls `add_actor`, and it discovers that actor requires `wascc:httpserver`. The custom host code could then call Gantry and ask for _any_ provider for `wascc:httpserver` that satisfies the triplet, and it gets the bytes and then just loads the provider. The inner workings of a "provider archive" would never be seen by clients/consumers of gantry.

The "user" never really sees the complexity, and the burden is really on capability provider developers that have to build for all the target platforms that might make sense. If that's just 1 platform, say all your deployments in production are x64 linux, then it's easy.

There are a couple of underlying facts/limitations with existing capability providers:
* You can't sign an ELF binary, nor can you embed arbitrary payloads inside it, so native capability providers right now are far less secure than their portable counterparts.
* Related to the first bullet, we have no revision data on capability providers (or any other metadata, for that matter)
* Again related to the first bullet, we don't have hashes on providers, so we can't prove they haven't been tampered with, which is a bigger security risk than actors because providers have unrestricted OS access

If you pre-compile the native capability provider into the host, then you sacrifice the ability for that host to choose at runtime which provider will satisfy the demand of the actors.

99% of the reason for the so-called `cpar` file is to be able to store a JWT alongside a corresponding capability provider. Storing per-architecture triplets is only the remaining 1%.

**Comment by autodidaddict** [comment]
> With the inclusion of statically linked capabilities, an easy way to use different capabilities regardless of architecture is available.

Statically linked providers all have the exact same architecture triplet as the host into which they've been compiled. Statically linked providers are inherently _less_ flexible than the dynamic ones because you can't swap static ones at runtime based on operations configuration / rules.

**Comment by autodidaddict** [comment]
The other thing to keep in mind is that the goal would be to throw away all of the triplet-specific stuff once WASI reaches a maturity level that all of our existing native providers can be compiled into WASI  `.wasm` files. My gut says we won't have stable WASI implementations like that for another  year, so we need some way to "deal with" native capability providers until then.

**Comment by brooksmtownsend** [comment]
> > With the inclusion of statically linked capabilities, an easy way to use different capabilities regardless of architecture is available.
> 
> Statically linked providers all have the exact same architecture triplet as the host into which they've been compiled. Statically linked providers are inherently _less_ flexible than the dynamic ones because you can't swap static ones at runtime based on operations configuration / rules.

So statically linked providers _feel_ flexible when using them, but at the end of the day don't improve the situation cross architectures, got it.

**Comment by brooksmtownsend** [comment]
In the event that the burden of development would be on us for Gantry, then I think it could be worthwhile so long as we keep a certain list of architectures that we support (could be a good inclusion to the Makefile, to build for all X architectures that we want to support). I was worried when, as a developer, requesting to add the httpserver capability and I'd have to do something like :
```
    host.add_native_capability(nats_capability, "x86_darwin_unknown");
```
Or whatever my architecture triplet ends up being.

**Comment by brooksmtownsend** [comment]
So what happens when revision data / versioning is implemented for capability providers? Will we give Gantry the power to choose an appropriate revision, or will the user be able to specify that they want a specific version of that capability provider?

**Comment by autodidaddict** [comment]
You'd still be able to add any capability you want, from any byte source you want. Today we don't have the ability to add capability providers from gantry... so you'd still do
```
host.add_native_capability(NativeCapability::from_file("foo.so")?;`
```

What I'm suggesting is that in the near future, you could _optionally_ source your capability providers from Gantry. If you did so, you would be responsible for telling the _host_ what architecture it is, so that the host would be able to tell Gantry under the hood, and you'd not see that happen in the following call:
```
host.add_capability_from_gantry("wascc:httpserver")?;
```
This would work because the host would've been configured with something like:
```
host.set_attribute("arch", "x86_64-unknown-linux-gnu");
```

If the host wanted to limit itself to only loading WASI capability providers, it could maybe do something like:
```
host.set_attribute("arch", "wasm32-wasi");
```
```

**Comment by autodidaddict** [comment]
Maybe:
```
host.add_from_gantry(EntityType::Provider, "wascc:httpserver", Some(12))?;
host.add_from_gantry(EntityType::Actor, "MABC...Z", None)?;
host.add_from_gantry(EntityType::Provider, "wascc:httpserver", None)?;
```

**Comment by brooksmtownsend** [comment]
Yeah okay, I like that flow then for getting things from Gantry, I think my view previously was limited as I've only really used Gantry as a middleman to get me something from my own disk anyways. It sounds like a good idea to get this configured at our own pace while we wait for WASI support, and then upgrade to the portable WASI modules when it's ready.

**Comment by autodidaddict** [comment]
Providers will need an identifier that allows for disambiguation of the implementation. For example, the NATS provider for `wascc:messaging` and a RabbitMQ provider for `wascc:messaging` must both be able to co-exist in Gantry _and_ must be able to be supplied to the waSCC host consumer in a way that lets the Gantry client select the appropriate implementation. I suggest a field like `implementation_id` or `implementor_id` or , possibly even better might be `vendor_id`
